### DISCUSSION TITLE
Attach Service Control Policy to AWS Organizations

### DISCUSSION THREAD

**Discussion by pete0emerson** [gruntwork-io/knowledge-base#258] [Help] [answered]
I am wondering how we may attach SCPs to OUs in the Gruntwork codebase.

My understanding is that we would need to use this module (https://github.com/gruntwork-io/terraform-aws-security/tree/v0.22.2/modules/aws-organizations), create the OU, create the SCP and attach an aws_organizations_policy (https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/organizations_policy). 

We also need to add SCP as an enabled policy type.

Any recommendations on how we might approach this?

**Comment by yorinasub17** [comment, marked as answer]
We currently don't offer a module to manage organizational units (OU) or SCPs, so what you have laid out in your understanding is correct: you will need to create a terraform module that uses the raw terraform resources to manage the OUs and SCPs.

However, this is a gap in our module catalog that we can address, so I have filed https://github.com/gruntwork-io/terraform-aws-security/issues/633 for us to consider implementing a dedicated module for managing OUs with child accounts and SCPs.

---

In the meantime, the best approach to implement this would be to create a module that manages the OU with SCP and which takes in the AWS Organization root ID to hook the OU to. The following is a skeleton that you can use:

```hcl
module "organization" {
  source = "git::git@github.com:gruntwork-io/module-security.git//modules/aws-organizations?ref=v0.62.3"

  organizations_feature_set = "ALL"
  organizations_enabled_policy_types = ["SERVICE_CONTROL_POLICY"] 

  # ... other args omitted for brevity ...
}

resource "aws_organizations_organizational_unit" "unit" {
  name = var.org_unit_name
  parent_id = module.organization.root_id
}

# Replicate for any accounts you wish to invite under the OU
resource "aws_organizations_account" "unit_child" {
  parent_id = aws_organizations_organizational_unit.unit.id
  # ... other args omitted for brevity ...
}

# Replicate the following for any SCPs you wish to attach to the OU
resource "aws_organizations_policy" "scp_policy" {
  name        = "Policy Name"
  description = "Policy description"

  content = data.aws_iam_policy_document.scp_policy.json
}

resource "aws_organizations_policy_attachment" "scp_policy" {
  policy_id = aws_organizations_policy.scp_policy.id
  target_id = aws_organizations_organizational_unit.unit.id
}

data "aws_iam_policy_document" "scp_policy" {
  # ... args omitted for brevity ...
}
```

**Reply by craigwong** [reply]
Hi @yorinasub17 Thanks for your answer. As we have a more involved (nested, 3-level) OU hierarchy, it sounds like creating our own module to house OUs and SCPs is the way to go. Thanks for the skeleton code and for filing the feature request--in the meantime, we will use your skeleton code, but do you expect this feature request to turn around in the next couple of months? Thank you.

**Reply by yorinasub17** [reply]
> do you expect this feature request to turn around in the next couple of months?

Unfortunately, I can't make any commitments on when we will implement feature requests. With that said, I do expect that we will be able to prioritize this in the next couple months, so I recommend subscribing to the ticket for updates.

**Reply by craigwong** [reply]
Ok sounds good. Thanks again!

**Comment by craigwong** [comment]
@yorinasub17 @pete0emerson Hi guys, sorry to contact you this way, but I'm unable to contact you through Slack due to Cisco only having the "free plan" on Slack.

We have a question regarding provider version pinning--how can we proceed?

**Reply by yorinasub17** [reply]
Hi Craig, you can use the same strategy indicated in https://github.com/gruntwork-io/knowledge-base/discussions/187, where you use an `_override.tf` file to force the provider versioning rules.

> Hi guys, sorry to contact you this way, but I'm unable to contact you through Slack due to Cisco only having the "free plan" on Slack.

No worries! In the future, it would be best to create a new discussion here with your question, or email support@gruntwork.io.

**Reply by craigwong** [reply]
Thanks @yorinasub17. Will do.
